test/sys/test_uio.rs

```diff
@@ -190,3 +190,51 @@ fn test_preadv() {
     let all = buffers.concat();
     assert_eq!(all, expected);
 }
+
+#[test]
+#[cfg(target_os = "linux")]
+// FIXME: qemu-user doesn't implement process_vm_readv/writev on most arches
+#[cfg_attr(not(any(target_arch = "x86", target_arch = "x86_64")), ignore)]
+fn test_process_vm_readv() {
+    use nix::unistd::ForkResult::*;
+    use nix::sys::signal::*;
+    use nix::sys::wait::*;
+    use std::str;
+
+    #[allow(unused_variables)]
+    let m = ::FORK_MTX.lock().expect("Mutex got poisoned by another test");
+
+    let (r, w) = pipe().unwrap();
+    match fork() {
+        Ok(Parent { child }) => {
+            close(w).unwrap();
+            let mut msg = vec![0u8; 32];
+            let bytes_read = read(r, &mut msg).unwrap();
+            msg.truncate(bytes_read);
+            close(r).unwrap();
+
+            let ptr: usize = str::from_utf8(&msg).unwrap().parse().unwrap();
+            let remote_iov = RemoteIoVec { base: ptr, len: 4 };
+            let mut buf = vec![0u8; 4];
+
+            let ret = process_vm_readv(child,
+                                       &[IoVec::from_mut_slice(&mut buf)],
+                                       &[remote_iov]);
+
+            kill(child, SIGTERM).unwrap();
+            waitpid(child, None).unwrap();
+
+            assert_eq!(Ok(4), ret);
+            assert_eq!(&buf, b"test");
+        },
+        Ok(Child) => {
+            let _ = close(r);
+            let s = String::from("test");
```

Comment:
Technically, you shouldn't use `String::from` in the child, because that requires allocating memory, and memory allocation isn't async signal safe.  The `format!` macro probably does allocation too.  See http://man7.org/linux/man-pages/man7/signal-safety.7.html .

Reply:
@asomers Let me know if this looks better. (I'm not actually sure if it's fully safe, in the sense that I'm not sure there's a guarantee that the writes to memory actually happen, but at least it's not async-signal-unsafe.)